File: templates/rangequery/fenwick.go
package rangequery

type Fenwick struct {
	tree []int
}

func NewFenwick(n int) *Fenwick {
	return &Fenwick{tree: make([]int, n+1)}
}

func (f *Fenwick) Add(index, delta int) {
	for i := index + 1; i < len(f.tree); i += i & -i {
		f.tree[i] += delta
	}
}

func (f *Fenwick) PrefixSum(index int) int {
	sum := 0
	for i := index + 1; i > 0; i -= i & -i {
		sum += f.tree[i]
	}
	return sum
}

func (f *Fenwick) RangeSum(left, right int) int {
	if left == 0 {
		return f.PrefixSum(right)
	}
	return f.PrefixSum(right) - f.PrefixSum(left-1)
}

File: templates/rangequery/fenwick.py
class Fenwick:
    def __init__(self, size: int) -> None:
        self.tree = [0] * (size + 1)

    def add(self, index: int, delta: int) -> None:
        i = index + 1
        while i < len(self.tree):
            self.tree[i] += delta
            i += i & -i

    def prefix_sum(self, index: int) -> int:
        total = 0
        i = index + 1
        while i > 0:
            total += self.tree[i]
            i -= i & -i
        return total

    def range_sum(self, left: int, right: int) -> int:
        if left == 0:
            return self.prefix_sum(right)
        return self.prefix_sum(right) - self.prefix_sum(left - 1)
